src/Chainweb/Pact/SPV.hs

```diff
@@ -169,7 +177,77 @@ verifySPV bdb bh typ proof = runExceptT $ go typ proof
 
       t -> throwError $! "unsupported SPV types: " <> t
 
-
+-- | A tag for specifying the format of base64 error messages on chain.
+--
+--   `Legacy` errors match those produced by our legacy version of
+--   base64-bytestring, and are produced by parsing the error messages
+--   we receive from our current version of base64-bytestring (1.2),
+--   and formatting them in the older style. Legacy behavior also implies
+--   that non-canonical encodings be allowed, because that was the behavior
+--   of the legacy bytestring parser; and improperly padded messages
+--   will get extra padding, because that is the legacy chainweb behavior.
+--
+--   `Simplified` errors are a static string, which may not describe
+--   the issue with the base64-encoded string as well, but will be
+--   more stable as we upgrade base64 decoding libraries in the future.
+--   In the `Simplified` errors setting, messages rejected for using
+--   non-canonical encodings will remain as errors, because we want to
+--   begin enforcing the expected constraints on base64-encoded messages.
+--   Similarly, Simplified parsing will not add extra padding to improperly
+--   padded messages.
+data GenerateBase64ErrorMessage
+  = Legacy
+  | Simplified
+
+
+-- | A modified version of `decodeBase64UrlNoPaddingText` that emits
+--   base64 decoding errors in a configurable way.
+decodeB64UrlNoPaddingTextWithFixedErrorMessage :: MonadThrow m => GenerateBase64ErrorMessage -> T.Text -> m B.ByteString
+decodeB64UrlNoPaddingTextWithFixedErrorMessage errorMessageType msg =
+  fromEitherM
+  . first (\e -> Base64DecodeException $ base64ErrorMessage e)
+  . (if errorMessageType == Legacy then patchNonCanonical else id)
+  . B64U.decode
+  . T.encodeUtf8
+  . (if errorMessageType == Legacy then pad else id)
+  $ msg
+  where
+    pad t = let s = T.length t `mod` 4 in t <> T.replicate ((4 - s) `mod` 4) "="
+    base64ErrorMessage m = case errorMessageType of
+      Legacy -> base64DowngradeErrorMessage (Text.pack m)
+      Simplified -> "could not base64-decode message"
+    patchNonCanonical decodeResult = case decodeResult of
+      Right bs -> Right bs
+      Left e | "non-canonical" `T.isInfixOf` T.pack e ->
+               Right (B64U.decodeLenient (T.encodeUtf8 msg))
```

Comment:
The reasoning behind this change in particular is that the previous base64-bytestring versions allowed "non-canonical" encodings in the sense that they are not true base64, but will decode to a particular string that will not round trip. Because we can assume that canonicity errors are the result of a structurally sound base64 input, we defer to the lenient decoder, which doesn't enforce canonicity. This is *terribly inefficient*, but since it only runs once or twice on a string of max size 1kb, we're not going to suffer too much.